larq/quantized_variable.py

```diff
@@ -0,0 +1,308 @@
+tf.register_tensor_conversion_function(
+    QuantizedVariable, QuantizedVariable._dense_var_to_tensor
+)
+ops.register_dense_tensor_like_type(QuantizedVariable)
+
+
+def create_quantized_variable(variable, quantizer=None, precision=None):
+    """Creates a QuantizedVariable that wraps another variable.
+
+    This typically just returns `QuantizedVariable(variable)`. But, if the variable
+    is a DistributedVariable or one of its subclasses, we instead dynamically
+    create a class that subclasses from both QuantizedVariable and
+    variable.__class__. This is so the returned variable will still pass
+    `isinstance(variable, variable.__class__)`, which is required for
+    DistributedVariables and its subclasses to work properly.
+
+    # Arguments
+    variable: A floating-point resource variable to wrap.
+    quantizer: An optional quantizer to transform the floating-point variable to a
+        fake quantized variable.
+    precision: An optional integer defining the precision of the quantized variable.
+        If `None`, `quantizer.precision` is used.
+
+    # Returns
+    A QuantizedVariable that wraps the variable.
+    """
+    if not isinstance(variable, DistributedVariable):  # type: ignore
+        return QuantizedVariable(variable, quantizer, precision)
+
+    class QuantizedDistributedVariable(QuantizedVariable, variable.__class__):
+        """A QuantizedVariable that also subclasses from DistributedVariable."""
+
+        def get(self, *args, **kwargs):
+            # For some reason this is needed to make unit `x + x` pass on TF 1.14
+            return self._quantize(self.latent_variable.get(*args, **kwargs))
+
+    return QuantizedDistributedVariable(variable, quantizer, precision)
+
+
+def _maybe_wrap(variable, quantizer, precision, wrap=True):
+    """Creates an QuantizedVariable that wraps another variable if applicable.
+
+    This function is used to wrap the return value of QuantizedVariable.assign.
+    Unfortunately MirroredVariable.assign will (incorrectly) return a Mirrored
+    value instead of a MirroredVariable. So we cannot properly wrap it in an
+    AutoCastVariable. We return the original variable in that case.
+
+    # Arguments
+    variable: A tf.Variable or op.
+    quantizer: An optional quantizer to transform the floating-point variable to a
+        fake quantized variable.
+    precision: An optional integer defining the precision of the quantized variable.
+        If `None`, `quantizer.precision` is used.
+    wrap: A boolean to define whether to wrap the variable in an QuantizedVariable.
+
+    # Returns
+    An QuantizedVariable if wrap is True and variable is a resource variable.
+    """
+    if wrap and resource_variable_ops.is_resource_variable(variable):
+        return create_quantized_variable(variable, quantizer, precision)
+    return variable
+
+
+def _get_name(obj):
+    try:
+        return obj.__name__
+    except AttributeError:
+        return obj.__class__.__name__
```

Comment:
Since this is just used in `QuantizedVariable`, I think it may be better to make it a static internal method of that class.

Reply:
:+1: Makes sense.